### ISSUE TITLE
Unexpected type error involving @KeyFor and covariant type parameter

### ISSUE THREAD

**Issue by GoogleCodeExporter** [typetools/checker-framework#173]
```
Assuming that @KeyFor({"map"}) T is a subtype of T and given that Map.Entry is 
covariant in its first type parameter (See 
http://code.google.com/p/checker-framework/source/browse/checkers/jdk/nullness/s
rc/java/util/Map.java#7), I expect that Entry<@KeyFor({"map"}) ? extends K, ? 
extends V> be a subtype of Entry<? extends K, ? extends V>. However, the 
checkers nullness processor (1.4.2) reports the following type error for the 
attached class:

MyMap.java:11: error: incompatible types in enhanced for loop.
    for (Map.Entry<? extends K, ? extends V> entry : map.entrySet()) {
                                                                 ^
  found   : Entry<@checkers.nullness.quals.KeyFor({"map"}) ? extends K extends Object, ? extends V extends Object>
  required: Entry<? extends K extends Object, ? extends V extends Object>
```

Original issue reported on code.google.com by `reprogra...@gmail.com` on 1 Nov 2012 at 6:58

Attachments:
- [MyMap.java](https://storage.googleapis.com/google-code-attachments/checker-framework/issue-173/comment-0/MyMap.java)
